### PR TITLE
Changes

### LINKED ISSUE #35: Swing timers are not updated directly when you gain Flurry, or when Flurry fades
The swing timer is only updated in swing_weapon() (Combat_simulator.cpp:1087), so Flurry gains are only applied on the next MH or OH swing, effectively hasting first one, then the other. If Flurry fades, the haste loss is handled in the same way.

In game, both gaining or losing Flurry immediately updates the swing timers for both weapons, e.g. for a 2.6 speed MH, if WW procs Flurry exactly mid-swing, the next MH hit is 1.3 + 1.3 / 125% = 2.34 seconds after the first. If Flurry faded 0.8 seconds after a swing, the next MH hit would be 0.8 + (2.6 / 125% - 0.8) * 125% = 2.4 seconds later.

### PR DESCRIPTION
**Pull request by vigo2**
Minor changes and bugfixes, including the flurry_haste_factor_ fix which might not cleanly apply.

"Just fixing" things is probably easier than filing a ton of minor issues. If this way of contributing is fine, I have a few larger changes in the queue.

For review, it's probably easiest to go commit-by-commit.

This fixes #35, #40, #41, and #42, plus a few non-filed issues (e.g. some of the "fishyness" mentioned in #32).

### CODE REVIEW

File: simulator/source/Combat_simulator.cpp [hunk 1]
```diff
@@ -1046,132 +1004,108 @@ void Combat_simulator::hit_effects(Weapon_sim& weapon, Weapon_sim& main_hand_wea
...
+        const auto& hit_outcome = hit_outcomes.emplace_back(generate_hit(main_hand_weapon, swing_damage, Hit_type::white, weapon.socket,
                                                special_stats, damage_sources));
-        if (dpr_heroic_strike_queued_)
-        {
-            rage -= heroic_strike_rage_cost;
-            dpr_heroic_strike_queued_ = false;
-        }
-        else if (dpr_cleave_queued_)
-        {
-            rage -= 20;
```
[the remark this answers is not in the document]
Comment (vigo2) [reply simulator/source/Combat_simulator.cpp:1144]: sample from the current master:

```
"normal" mode
white (mh)    = 250.979
white (oh)    = 89.2602
heroic strike = 145.866
bloodthirst   = 161.346
whirlwind     = 90.2794
execute       = 107.194
----------------------
total         = 844.926

"dpr (heroic strike)" mode
white (mh)    = 341.14
white (oh)    = 85.3679
heroic strike = 0
bloodthirst   = 155.57
whirlwind     = 86.6012
execute       = 105.961
----------------------
total         = 774.639
```

BT/WW/Execute damage is /lower/ in "dpr mode", even though there should be additional rage (the re-gained white MH rage) available.

File: simulator/source/Combat_simulator.cpp [hunk 2]
```diff
@@ -1911,19 +1766,31 @@ void Combat_simulator::simulate(const Character& character, int init_iteration,
...
+            }
+
+            if (oh_swing)
+            {
+                weapons[1].internal_swing_timer = weapons[1].swing_speed / (1 + special_stats.haste);
+            }
+            else if (is_dual_wield && special_stats.haste != oldHaste)
+            {
+                weapons[1].internal_swing_timer *= (1 + oldHaste) / (1 + special_stats.haste);
+            }
```
Comment (vigo2) [inline simulator/source/Combat_simulator.cpp:1793]: speed up or slow down remaining swing timers, or reset with updated haste. this is necessary for flurry, but also for any other haste proc.

**Review by TheGroxEmpire — COMMENTED** [review]
Thank you so much for this fix and changes. I have tested and reviewed the code, they look good to me!
Before I approved this and merging it into the master, could you please list the changes in a list for patch notes? You can take example from what I did on previous update patch notes.

**Comment by vigo2** [comment]
Something like

- Deep wounds damage was dealing 100% instead of up to 60% weapon damage
- Dodged or missed executes no longer consume remaining rage
- Flurry now only grants 25% instead of 30% haste
- Flurry uptime is now, in fact, the uptime of flurry haste, and no longer the percentage of main- or off-hand swings while flurry is active
- Haste changes now apply immediately to both weapons
- Fixed a bug that was granting a small amount of additional rage on white dodges and misses
- Heroic Strike and Cleave were overvalued for DPR calculation
- Off-hand damage was accidentally almost halved. It's now back to where it belongs ;)

?

**Review by TheGroxEmpire — APPROVED** [review]
Yep that seems right.

**Pull request merged**